python/pyspark/sql/pandas/conversion.py

```diff
@@ -0,0 +1,431 @@
+                dtype[field.name] = np.object
+
+        for f, t in dtype.items():
+            pdf[f] = pdf[f].astype(t, copy=False)
+
+        if timezone is None:
+            return pdf
+        else:
+            from pyspark.sql.pandas.types import _check_series_convert_timestamps_local_tz
+            for field in self.schema:
+                # TODO: handle nested timestamps, such as ArrayType(TimestampType())?
+                if isinstance(field.dataType, TimestampType):
+                    pdf[field.name] = \
+                        _check_series_convert_timestamps_local_tz(pdf[field.name], timezone)
+            return pdf
+
+    @staticmethod
+    def _to_corrected_pandas_type(dt):
+        """
+        When converting Spark SQL records to Pandas :class:`DataFrame`, the inferred data type
+        may be wrong. This method gets the corrected data type for Pandas if that type may be
+        inferred incorrectly.
+        """
+        import numpy as np
+        if type(dt) == ByteType:
+            return np.int8
+        elif type(dt) == ShortType:
+            return np.int16
+        elif type(dt) == IntegerType:
+            return np.int32
+        elif type(dt) == LongType:
+            return np.int64
+        elif type(dt) == FloatType:
+            return np.float32
+        elif type(dt) == DoubleType:
+            return np.float64
+        elif type(dt) == BooleanType:
+            return np.bool
+        elif type(dt) == TimestampType:
+            return np.datetime64
+        else:
+            return None
+
+    def _collect_as_arrow(self):
+        """
+        Returns all records as a list of ArrowRecordBatches, pyarrow must be installed
+        and available on driver and worker Python environments.
+
+        .. note:: Experimental.
+        """
+        from pyspark.sql.dataframe import DataFrame
+
+        assert isinstance(self, DataFrame)
+
+        with SCCallSiteSync(self._sc):
+            port, auth_secret, jsocket_auth_server = self._jdf.collectAsArrowToPython()
+
+        # Collect list of un-ordered batches where last element is a list of correct order indices
+        try:
+            results = list(_load_from_socket((port, auth_secret), ArrowCollectSerializer()))
+        finally:
+            # Join serving thread and raise any exceptions from collectAsArrowToPython
+            jsocket_auth_server.getResult()
+
+        # Separate RecordBatches from batch order indices in results
+        batches = results[:-1]
+        batch_order = results[-1]
+
+        # Re-order the batch list using the correct order
+        return [batches[i] for i in batch_order]
+
+
+class SparkConversionMixin(object):
+    """
+    Min-in for the conversion from pandas to Spark. Currently, only :class:`SparkSession`
+    can use this class.
+    """
+    def createDataFrame(self, data, schema=None, samplingRatio=None, verifySchema=True):
```

Comment:
Instead of shading `SparkSession.createDataFrame` and later [having to disambiguate](https://github.com/HyukjinKwon/spark/blob/c63c271d95e7864b69ca477c51e4e2d22b7d9029/python/pyspark/sql/session.py#L600-L601) it would be more idiomatic to just name mangle [like this](https://github.com/zero323/spark/blob/228c1f3896f871aae303eb306d6233bcf28edf19/python/pyspark/sql/pandas/conversion.py#L226)

```python
def __createDataFrame(self, data, schema=None, samplingRatio=None, verifySchema=True):
```

[and later](https://github.com/zero323/spark/blob/228c1f3896f871aae303eb306d6233bcf28edf19/python/pyspark/sql/session.py#L600)

```python
return self._SparkConversionMixin__createDataFrame(
```

It would be also easier to add new mixins later, if that proves to be useful for some reason ([here is possible diff](https://github.com/apache/spark/compare/master...zero323:SPARK-30434-mangle?expand=1)).

Additionally we could limit  the responsibility to data and schema preparation only ([diff for this](https://github.com/apache/spark/compare/master...zero323:SPARK-30434-prepare?expand=1)). Additionally to benefits mentioned above, it allow us to decouple mix in from `SparkSession` (as we'd only need `_wrapped._conf`).

Reply:
`_SparkConversionMixin__createDataFrame` looks fine in a way that it can distinguish which mixin is used but we don't really use mangle name yet in the current codebase too.

Reply:
> I did the current way so that SparkConversionMixin.createDataFrame can work alone by itself

But I cannot, can it? It still has to be mixed-in into something that has these methods.

>  I didn't like the second way because it's weird to return both schema and data.

Matter of taste I guess.

> _SparkConversionMixin__createDataFrame looks fine in a way that it can distinguish which mixin is used but we don't really use mangle name yet in the current codebase too.

That's true, though I am not aware of  case might occur. We use mixins for `Params`, but there conflicts are either unlikely or irrelevant. Here both methods are used, and it is theoretically possible that other similar might follow. Let's say `pypark.sql.ray.conversions` or ``pypark.sql.dask.conversions`, either here, or in the 3rd party extensions. In such cases we might prefer to have consistent interface.

Anyway, I don't have very strong opinion about this. I've mentioned this primarily because corresponding type hints looked funky.

There is however  one  immediate side effect of the current approach ‒ it  makes passing methods values a bit unreliable (and by extension methods that depend on it, like monkey patching).